File: tests/test-minecraft-osp/src/main/java/pize/tests/voxelgame/main/entity/Entity.java
package pize.tests.voxelgame.main.entity;

import pize.math.Maths;
import pize.math.util.EulerAngles;
import pize.math.vecmath.vector.Vec3f;
import pize.physic.BoundingBox3;
import pize.physic.BoxBody;
import pize.physic.Collider3f;
import pize.physic.Velocity3f;
import pize.tests.voxelgame.main.Direction;
import pize.tests.voxelgame.client.block.BlockProperties;
import pize.tests.voxelgame.client.block.shape.BlockCollide;
import pize.tests.voxelgame.main.Tickable;
import pize.tests.voxelgame.main.block.BlockData;
import pize.tests.voxelgame.main.chunk.ChunkUtils;
import pize.tests.voxelgame.main.level.Level;

import java.util.ArrayList;
import java.util.UUID;

public abstract class Entity extends BoxBody implements Tickable{
    
    private Level level;
    private final EntityType<?> entityType;
    private final EulerAngles rotation;
    private final Velocity3f velocity;
    private UUID uuid;
    
    private BoxBody[] blockBoxes;
    private boolean onGround;
    
    public Entity(EntityType<?> entityType, Level level){
        super(entityType.getBoundingBox());
        
        this.level = level;
        this.entityType = entityType;
        this.rotation = new EulerAngles();
        this.velocity = new Velocity3f();
        this.uuid = UUID.randomUUID();
    }
    
    
    public Level getLevel(){
        return level;
    }
    
    public void setLevel(Level level){
        this.level = level;
    }
    
    public EntityType<?> getEntityType(){
        return entityType;
    }
    
    public EulerAngles getRotation(){
        return rotation;
    }
    
    public Velocity3f getVelocity(){
        return velocity;
    }
    
    
    public UUID getUUID(){
        return uuid;
    }
    
    public void setUUID(UUID uuid){
        this.uuid = uuid;
    }
    
    public float getEyeHeight(){
        return getBoundingBox().getSizeY() * 0.85F;
    }
    

    @Override
    public void tick(){
        // Check is chunk loaded
        final Vec3f pos = getPosition();
        if(level.getBlockChunk(pos.xf(), pos.zf()) == null)
            return;
        
        // Update blocks around player
        blockBoxes = getBlockBoxes();
        // Update is player on ground

        onGround = isCollidedTo(Direction.NEGATIVE_Y);
    }
    
    public boolean isOnGround(){
        return onGround;
    }
    
    public boolean isCollidedTo(Direction face){
        final Vec3f dir = new Vec3f(face.getNormal()).mul(Maths.Epsilon);
        return Collider3f.getCollidedMotion(this, dir, blockBoxes).len2() < dir.len2();
    }
    
    public boolean isCollidedTo(Vec3f direction){
        final Vec3f dir = direction.copy().nor().mul(Maths.Epsilon);
        return Collider3f.getCollidedMotion(this, dir, blockBoxes).len2() < dir.len2();
    }
    
    /** Get Bounding Boxes of blocks around Entity */
    private BoxBody[] getBlockBoxes(){
        final ArrayList<BoxBody> blockBoxes = new ArrayList<>();
        
        final Velocity3f velocity = getVelocity();
        final Vec3f min = this.getMin();
        final Vec3f max = this.getMax();
        
        final int beginX = Maths.floor(min.x - 0.5 + Math.min(0, velocity.x));
        final int beginY = Math.max(0, Math.min(ChunkUtils.HEIGHT_IDX,
            Maths.floor(min.y - 0.5 + Math.min(0, velocity.y))
        ));
        final int beginZ = Maths.floor(min.z - 0.5 + Math.min(0, velocity.z));
        
        final int endX = Maths.ceil(max.x + 0.5 + Math.max(0, velocity.x));
        final int endY = Math.max(0, Math.min(ChunkUtils.HEIGHT,
            Maths.ceil(max.y + 0.5 + Math.max(0, velocity.y))
        ));
        final int endZ = Maths.ceil(max.z + 0.5 + Math.max(0, velocity.z));
        
        for(int x = beginX; x < endX; x++)
            for(int y = beginY; y < endY; y++)
                for(int z = beginZ; z < endZ; z++){

                    final short blockData = level.getBlock(x, y, z);
                    final byte blockState = BlockData.getState(blockData);
                    final BlockProperties block = BlockData.getProps(blockData);

                    if(block.getStates().isEmpty())
                        continue;

                    final BlockCollide shape = block.getState(blockState).getCollide();
                    if(shape == null)
                        continue;
                    
                    for(BoundingBox3 boundingBox: shape.getBoxes()){
                        final BoxBody box = new BoxBody(boundingBox);
                        box.getPosition().set(x, y, z);
                        
                        blockBoxes.add(box);
                    }
                }
        
        return blockBoxes.toArray(new BoxBody[0]);
    }
    
    protected Vec3f moveEntity(Vec3f motion){
        if(blockBoxes == null)
            return null;
        
        final Vec3f collidedMove = Collider3f.getCollidedMotion(this, motion, blockBoxes);
        getPosition().add(collidedMove);
        
        return collidedMove;
    }
    
}

File: tests/test-minecraft-osp/src/main/kotlin/pize/tests/voxelgame/main/entity/Entity.kt
package pize.tests.voxelgame.main.entity

import pize.graphics.camera.Camera3D.x
import pize.graphics.camera.Camera3D.y
import pize.graphics.camera.Camera3D.z
import pize.math.Maths
import pize.math.Maths.ceil
import pize.math.Maths.floor
import pize.math.util.EulerAngles
import pize.math.vecmath.vector.Vec3f
import pize.physic.BoxBody
import pize.physic.Collider3f.getCollidedMotion
import pize.physic.Velocity3f
import pize.tests.voxelgame.main.Direction
import pize.tests.voxelgame.main.Tickable
import pize.tests.voxelgame.main.block.BlockData
import pize.tests.voxelgame.main.chunk.ChunkUtils
import pize.tests.voxelgame.main.level.Level
import java.util.*
import kotlin.math.max
import kotlin.math.min

abstract class Entity(val entityType: EntityType<*>, var level: Level?) : BoxBody(
    entityType.boundingBox
), Tickable {
    val rotation: EulerAngles
    val velocity: Velocity3f
    var uUID: UUID?
    private var blockBoxes: Array<BoxBody>?
    var isOnGround = false
        private set

    init {
        rotation = EulerAngles()
        velocity = Velocity3f()
        uUID = UUID.randomUUID()
    }

    open val eyeHeight: Float
        get() = boundingBox!!.sizeY * 0.85f

    override fun tick() {
        // Check is chunk loaded
        val pos = position
        if (level!!.getBlockChunk(pos.xf(), pos.zf()) == null) return

        // Update blocks around player
        blockBoxes = getBlockBoxes()
        // Update is player on ground
        isOnGround = isCollidedTo(Direction.NEGATIVE_Y)
    }

    fun isCollidedTo(face: Direction): Boolean {
        val dir = Vec3f(face.normal).mul(Maths.Epsilon)
        return getCollidedMotion(this, dir, *blockBoxes!!).len2() < dir.len2()
    }

    fun isCollidedTo(direction: Vec3f): Boolean {
        val dir = direction.copy().nor().mul(Maths.Epsilon)
        return getCollidedMotion(this, dir, *blockBoxes!!).len2() < dir.len2()
    }

    /** Get Bounding Boxes of blocks around Entity  */
    private fun getBlockBoxes(): Array<BoxBody> {
        val blockBoxes = ArrayList<BoxBody>()
        val velocity = velocity
        val min = min
        val max = max
        val beginX = floor(min.x - 0.5 + min(0.0, velocity.x.toDouble()))
        val beginY = max(
            0.0, min(
                ChunkUtils.HEIGHT_IDX.toDouble(),
                floor(min.y - 0.5 + min(0.0, velocity.y.toDouble()))
                    .toDouble()
            )
        ).toInt()
        val beginZ = floor(min.z - 0.5 + min(0.0, velocity.z.toDouble()))
        val endX = ceil(max.x + 0.5 + max(0.0, velocity.x.toDouble()))
        val endY = max(
            0.0, min(
                ChunkUtils.HEIGHT.toDouble(),
                ceil(max.y + 0.5 + max(0.0, velocity.y.toDouble()))
                    .toDouble()
            )
        ).toInt()
        val endZ = ceil(max.z + 0.5 + max(0.0, velocity.z.toDouble()))
        for (x in beginX until endX) for (y in beginY until endY) for (z in beginZ until endZ) {
            val blockData = level!!.getBlock(x, y, z)
            val blockState = BlockData.getState(blockData)
            val block = BlockData.getProps(blockData)
            if (block.states.isEmpty()) continue
            val shape = block!!.getState(blockState.toInt()).collide ?: continue
            for (boundingBox in shape.boxes) {
                val box = BoxBody(boundingBox)
                box.position[x, y] = z
                blockBoxes.add(box)
            }
        }
        return blockBoxes.toTypedArray<BoxBody>()
    }

    protected fun moveEntity(motion: Vec3f?): Vec3f? {
        if (blockBoxes == null) return null
        val collidedMove = getCollidedMotion(this, motion!!, *blockBoxes!!)
        position.add(collidedMove)
        return collidedMove
    }
}
